src/otx/engine/engine.py

```diff
@@ -804,6 +808,14 @@ def _build_trainer(self, **kwargs) -> None:
         """Instantiate the trainer based on the model parameters."""
         if self._cache.requires_update(**kwargs) or self._trainer is None:
             self._cache.update(**kwargs)
+            # set up xpu device
+            if self._device.accelerator == DeviceType.xpu:
+                self._cache.update(strategy="xpu_single")
```

Comment:
Why is this needed? Is there any problem for just giving `accelerator==xpu`?

Reply:
After looking at `xpu_precision.py` and `xpu_single.py`, I see the situation. However, my gut tells me that `SingleDeviceStrategy` and `MixedPrecisionPlugin` must be device-independent. That means this combination `accelerator=xpu`, `precision=16`, and `strategy=single` should be working. Can we re-register them with patching XPU support rather than creating explicit ones? If it is hard to be done in this PR, please create another one.

Reply:
Patching strategy seems to be overwhelming. I tried to do that, it leads to errors and make code less readable in my opinion. Registering new strategy is ordinary process for new device. In lightning they have new strategy for each device.
Also, please, be aware of this draft (for a while) PR from lightning: https://github.com/Lightning-AI/pytorch-lightning/pull/17700

When they decide to actually add XPU as new device, we can remove everything and use XPU natively. Let's keep explicit strategy registering for now. I don't see any problems with that, it is how lightning works with new devices.

Reply:
> Registering new strategy is ordinary process for new device. In lightning they have new strategy for each device.

I visited the link you gave https://github.com/Lightning-AI/pytorch-lightning/pull/17700. However, it could not see the new implementation about the strategy. On the other hand, there are modifications for the existing strategy.
![image](https://github.com/openvinotoolkit/training_extensions/assets/26541465/439cd2fa-7d87-48a2-a225-99fe0063b0d5)

> When they decide to actually add XPU as new device, we can remove everything and use XPU natively. Let's keep explicit strategy registering for now.

1. I don't know whether the upstream situation is resolved near future. This is because https://github.com/Lightning-AI/pytorch-lightning/pull/17700 was opened almost a year ago and is not merged even today.
2. Even if the upstream PR is merged, I can expect that we still need some customizations for us like https://github.com/openvinotoolkit/training_extensions/blob/98e0b69e2903fa1d4ce799c1ee5480416f52b0ee/src/otx/algo/detection/utils/mmcv_patched_ops.py
It means the re-registering will exist in the future as well.

> Patching strategy seems to be overwhelming. I tried to do that, it leads to errors and make code less readable in my opinion.

I don't know why it is hard actually. I think it can be

```python
from lightning.pytorch.strategies.single_device import SingleDeviceStrategy as _SingleDeviceStrategy

class SingleDeviceStrategy(_SingleDeviceStrategy):
    ....

    def setup_optimizers(self, trainer: pl.Trainer) -> None:
        if self.root_device.type == "xpu":
            return self._xpu_setup_optimizer(trainer)
        return super().setup_optimizers(trainer)

    def _xpu_setup_optimizers(self, trainer: pl.Trainer) -> None:
        """Sets up optimizers."""
        super().setup_optimizers(trainer)
        if len(self.optimizers) != 1:  # type: ignore[has-type]
            msg = "XPU strategy doesn't support multiple optimizers"
            raise RuntimeError(msg)
        if trainer.task != "SEMANTIC_SEGMENTATION":
            model, optimizer = torch.xpu.optimize(trainer.model, optimizer=self.optimizers[0])  # type: ignore[has-type]
            self.optimizers = [optimizer]
            self.model = model
```
I can see nothing that can be error-prone.

Reply:
Let's consider it in the next PR